back/engines/commercial/multi_tenancy/lib/tasks/core/migrate_flexible_input_forms.rake

```diff
@@ -0,0 +1,170 @@
+# frozen_string_literal: true
+
+# to persist changes run: fix_existing_tenants:migrate_flexible_input_forms[true]
+# to persist changes for one host run: fix_existing_tenants:migrate_flexible_input_forms[true,localhost]
+namespace :fix_existing_tenants do
+  desc 'Migrate ideation form custom fields to the new codes and types'
+  task :migrate_flexible_input_forms, %i[persist_changes specify_host] => [:environment] do |_t, args|
+    persist_changes = args[:persist_changes] == 'true'
+    specify_host = args[:specify_host]
+    Rails.logger.info 'DRY RUN: Changes will not be persisted' unless persist_changes
+    stats = {}
+    Tenant.creation_finalized.each do |tenant|
+      next unless tenant.host == specify_host || specify_host.blank?
+
+      Rails.logger.info "PROCESSING TENANT: #{tenant.host}..."
+      Apartment::Tenant.switch(tenant.schema_name) do
+        form_migrator = FlexibleInputFormMigrator.new
+        CustomForm.all.each do |custom_form|
+          form_migrator.migrate_form(custom_form, persist_changes)
+        end
+        stats[tenant.host] = form_migrator.stats
+      end
+    end
+
+    stats.each do |host, stat|
+      Rails.logger.info "STATS: #{host} - forms:#{stat[:forms]}(#{stat[:empty]}) update:#{stat[:updated]} create:#{stat[:created]}"
+      stat[:errors].each do |error|
+        Rails.logger.info "ERROR: #{error}"
+      end
+    end
+  end
+end
+
+class FlexibleInputFormMigrator
+  def initialize
+    @stats = { forms: 0, empty: 0, updated: 0, created: 0, errors: [] }
+    @errors = []
+  end
+
+  attr_reader :stats
+
+  def migrate_form(custom_form, persist_changes)
+    @stats[:forms] += 1
+
+    # Get persisted fields by order
+    fields = CustomField.where(resource: custom_form).order(:ordering)
```

Comment:
I don't know to what extend we should be concerned about the execution time, I think it might be acceptable. But I think we could include the custom fields in the custom forms to avoid this query for each form. I think in most cases there are no persisted fields? That would mean that no SQL query would take place for all these forms.

Reply:
Yeah that works. Have fixed.